### PR TITLE
Better memory protections

### PR DESCRIPTION
**Pull request by Restioson**
Checklist:
- [x] Verified that the guard page works
- [x] Higher half
- [x] Map readonly sections of the kernel as readonly, etc
- [x] Clean up higher half, etc - in progress
- [x] Use an IST for interrupts
- [x] Guard pages for IST stacks
- [x] Correctly unmap heap pages

### CODE REVIEW

**Comment by Restioson** [comment]
@gegy1000 right, this is feature-complete now. Could you update jenkins rust version and see if that fixes the compile error?

**Comment by Restioson** [comment]
Forgot that I have two things left to do -- they have been added to the list.

**Comment by Restioson** [comment]
Everything is done now.

File: kernel/src/drivers/vga.rs [hunk 1]
```diff
@@ -30,7 +33,7 @@ impl fmt::Debug for VgaWriter {
 impl VgaWriter {
     pub const fn new() -> Self {
         VgaWriter {
-            buffer: unsafe { Unique::new_unchecked(0xb8000 as *mut _) },
+            buffer: unsafe { Unique::new_unchecked((VIRTUAL_VGA_PTR) as *mut _) },
```
Comment (gegy1000) [inline kernel/src/drivers/vga.rs]: Parenthesis can be elided here

File: kernel/src/interrupts/legacy_pic.rs [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    inner: [Pic; 2],
+}
+
+impl ChainedPics {
+    pub const fn new(offsets: (u8, u8)) -> Self {
+        unsafe {
+            ChainedPics {
+                inner: [
+                    Pic::new(offsets.0,
+                             SynchronizedPort::new(0x20),
```
Comment (gegy1000) [inline kernel/src/interrupts/legacy_pic.rs]: Formatting should probably be consistent with below

File: kernel/src/memory/buddy_allocator.rs [hunk 3]
```diff
@@ -92,14 +93,15 @@ macro_rules! buddy_allocator_bitmap_tree {
                     block_begin += 1 << ($BASE_ORDER);
                 }
 
-                let mut start: usize = 1 << (MAX_ORDER - 1);
-                for level in (0..MAX_ORDER).rev() {
-                    for node_index in start..(start +  blocks_in_level(level)) {
-                        tree.update_block(node_index, level);
+                let mut start: usize = 1 << (MAX_ORDER - 1) ;
```
Comment (gegy1000) [inline kernel/src/memory/buddy_allocator.rs]: Formatting weirdness at the end of the line here

File: kernel/src/memory/buddy_allocator.rs [hunk 4]
```diff
@@ -190,49 +190,28 @@ macro_rules! buddy_allocator_bitmap_tree {
...
                 assert!(index < BLOCKS_IN_TREE, "Block index {} out of bounds!", index);
+
                 assert_eq!(
                     unsafe { self.block(index - 1).order_free },
                     0,
                     "Block to free (index {}) must be used!",
                     index,
                 );
 
-                // Only if order isn't 0 we need to check the children, as blocks of order 0 have
```
Comment (gegy1000) [inline kernel/src/memory/buddy_allocator.rs]: Why was this removed, btw?

File: kernel/src/memory/heap.rs [hunk 5]
```diff
@@ -144,15 +168,30 @@ impl Heap {
 
             PAGE_TABLES.lock().unmap(
                 Page::containing_address(page_addr, PageSize::Kib4),
-                false,
+                false, // Do not free backing memory
```
Comment (gegy1000) [inline kernel/src/memory/heap.rs]: Would be nice to use enums here, would remove the need for booleans

File: kernel/src/memory/heap.rs [hunk 6]
```diff
@@ -144,15 +168,30 @@ impl Heap {
...
+    }
+
+    pub fn is_free(&self, ptr: *const u8, layout: Layout) {
+        let order = order(layout.size());
+        let global_ptr = ptr;
+        let ptr = ptr as usize - HEAP_START;
+
+        let level = MAX_ORDER - order;
+        let level_offset = super::buddy_allocator::blocks_in_level(level);
+        let index = level_offset + ((ptr as usize) >> (order + 6));
```
Comment (gegy1000) [inline kernel/src/memory/heap.rs]: 6 seems pretty arbitrary -- maybe extract to constant?

File: kernel/src/memory/paging/mod.rs [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+        const GLOBAL = 1 << 8;
+        /// Do not allow executing code from this page. NXE bit in EFER must be set.
+        const NO_EXECUTE = 1 << 63;
+    }
+}
+
+/// A trait that indicates a type represents a page table level
+pub trait TableLevel {}
+
+pub enum Level4 {}
```
Comment (gegy1000) [inline kernel/src/memory/paging/mod.rs]: Do enums get their index optimized away when no values?

File: kernel/src/interrupts/legacy_pic.rs (hunk 2 above)
Comment (Restioson) [reply kernel/src/interrupts/legacy_pic.rs]: The inclusion of this file in actually an accident.

File: kernel/src/memory/buddy_allocator.rs (hunk 4 above)
Comment (Restioson) [reply kernel/src/memory/buddy_allocator.rs]: The whole algorithm was buggy... idk. It works now and all tests pass lol.

File: kernel/src/memory/heap.rs (hunk 5 above)
Comment (Restioson) [reply kernel/src/memory/heap.rs]: They are booleans though?

File: kernel/src/memory/heap.rs (hunk 6 above)
Comment (Restioson) [reply kernel/src/memory/heap.rs]: 6 is the base order. Will extract though

File: kernel/src/memory/paging/mod.rs (hunk 7 above)
Comment (Restioson) [reply kernel/src/memory/paging/mod.rs]: Enums like this cannot ever exist as they have no values. This is never constructed, it is just a type level thing.

File: kernel/src/memory/physical_allocator.rs [hunk 8]
```diff
@@ -185,6 +182,19 @@ impl<'a> PhysicalAllocator<'a> {
 
         tree.deallocate(local_ptr, order);
     }
+
+    pub fn is_free(&self, ptr: *const u8, order: u8) {
```
[the remark this answers is not in the document]
Comment (gegy1000) [reply kernel/src/memory/physical_allocator.rs]: oops, meant to submit this review earlier

[the remark this answers is not in the document]
Comment (Restioson) [reply kernel/src/memory/physical_allocator.rs]: Nope... i believe it was debugging stuff
